pkg/dns/dns.go

```diff
@@ -141,29 +141,16 @@ func NewKubeDNS(client clientset.Interface, domain string, federations map[strin
 func (kd *KubeDNS) Start() {
 	go kd.endpointsController.Run(wait.NeverStop)
 	go kd.serviceController.Run(wait.NeverStop)
-	// Wait synchronously for the Kubernetes service and add a DNS record for it.
-	// This ensures that the Start function returns only after having received Service objects
-	// from APIServer.
-	// TODO: we might not have to wait for kubernetes service specifically. We should just wait
-	// for a list operation to be complete from APIServer.
-	kd.waitForKubernetesService()
-}
-
-func (kd *KubeDNS) waitForKubernetesService() (svc *kapi.Service) {
-	name := fmt.Sprintf("%v/%v", kapi.NamespaceDefault, kubernetesSvcName)
-	glog.Infof("Waiting for service: %v", name)
-	var err error
-	servicePollInterval := 1 * time.Second
+	// wait for both controllers have completed an initial resource listing
+	tick := time.Tick(500 * time.Millisecond)
 	for {
-		svc, err = kd.kubeClient.Core().Services(kapi.NamespaceDefault).Get(kubernetesSvcName)
-		if err != nil || svc == nil {
-			glog.Infof("Ignoring error while waiting for service %v: %v. Sleeping %v before retrying.", name, err, servicePollInterval)
-			time.Sleep(servicePollInterval)
-			continue
+		<-tick
+		if kd.endpointsController.HasSynced() && kd.serviceController.HasSynced() {
+			glog.V(0).Infof("Initialized services and endpoints from apiserver")
+			return
 		}
-		break
+		glog.V(0).Infof("DNS server not ready, retry in 500 milliseconds")
```

Comment:
This could spam a lot to the log.  Is there a practical timeout to this - maybe 60 seconds then abort the process?